nixos/modules/services/web-servers/h2o/default.nix

```diff
@@ -258,6 +384,67 @@ in
           ${lib.getExe cfg.package} ${lib.strings.escapeShellArgs args}
         '';
     };
-  };
 
+    # This service waits for all certificates to be available before reloading
+    # H2O configuration. `tlsTargets` are added to `wantedBy` + `before` which
+    # allows the `acme-finished-$cert.target` to signify the successful updating
+    # of certs end-to-end.
+    systemd.services.h2o-config-restart =
+      let
+        tlsTargets = map (certName: "acme-${certName}.target") certNames.all;
+        tlsServices = map (certName: "acme-${certName}.service") certNames.all;
+      in
+      mkIf (certNames.all != [ ]) {
+        wantedBy = tlsServices ++ [ "multi-user.target" ];
+        before = tlsTargets;
+        after = tlsServices;
+        unitConfig = {
+          ConditionPathExists = lib.optionals (tlsServices != [ ]) (
```

Comment:
At this point, is it possible for `tlsServices` to be `[]`? IIUC, the `mkIf` above would prevent that. Perhaps this could be an assert instead?